### PR TITLE
feat(ingest): allow extracting snowflake tags

### PR DESCRIPTION
**Pull request by frsann**
Adding support to extracts tags object tags from Snowflake. The scope is limited to database, schema, table/view, and column tags. It also (optionally) creates the extracted tags. Only the tags that are applied to objects are created; it does not pull all existing tags in Snowflake. Caters to https://feature-requests.datahubproject.io/p/ingest-snowflake-object-tags

There are some peculiarities about Snowflake tags that should be considered.

### 1. Snowflake tags are key-values, not simple strings (as in Datahub).

This is handled in this PR by converting the key-value tag into a simple string with the following formatting: `<db_name>.<schema_name>.<tag_name>=<tag_value>`, where `db_name` and `schema_name` are the DB and schema in which the tag is defined, respectively. While not the prettiest, this format should allow users to quite easily define a formatter to customize the look of the tag, if they chose to. Implementing this formatter is outside the scope of this PR.  

### 2. Snowflake allows tagging objects directly, but also through propagation (or lineage as they call it). 

This means that a tag applied to a database propagates down to the schemas, tables, and columns as well. Applying the same, propagated tag directly on the object, but with a different value, takes precedence over the propagated value. There are different methodologies to get the tags that are directly applied and the ones that are also propagated. 

As the propagated tags are "effective" (for example tag_based masking policies applied to tables evaluated when querying columns), a Datahub user might want extract them in addition to the directly applied tags. To accommodate both uses cases, a configuration field `extract_tags` is introduced that allows selecting which type of tags extraction is wanted.

It should be noted that extracting the propagated tags is done on an object basis (except for columns) and will therefore require more queries to be made. Extracting only the directly applied tags can be done in a single query per DB (+ cached).  

## Checklist

- [x] The PR conforms to DataHub's [Contributing Guideline](https://github.com/datahub-project/datahub/blob/master/docs/CONTRIBUTING.md) (particularly [Commit Message Format](https://github.com/datahub-project/datahub/blob/master/docs/CONTRIBUTING.md#commit-message-format))
- [x] Links to related issues (if applicable)
- [x] Tests for the changes have been added/updated (if applicable)
- [x] Docs related to the changes have been added/updated (if applicable). If a new feature has been added a Usage Guide has been added for the same.
- [x] For any breaking change/potential downtime/deprecation/big changes an entry has been made in [Updating DataHub](https://github.com/datahub-project/datahub/blob/master/docs/how/updating-datahub.md)

### CODE REVIEW

File: metadata-ingestion/src/datahub/ingestion/source/snowflake/snowflake_schema.py [hunk 1]
```diff
@@ -27,6 +28,17 @@ class SnowflakeFK:
...
 
+@dataclass
+class SnowflakeTag:
+    database: str
+    schema: str
+    name: str
+    value: str
+
+    def __str__(self):
+        return f"{self.database}.{self.schema}.{self.name}={self.value}".lower()
```
Comment (frsann) [inline metadata-ingestion/src/datahub/ingestion/source/snowflake/snowflake_schema.py:39]: The format of the stringification can be discussed.

File: metadata-ingestion/src/datahub/ingestion/source/snowflake/snowflake_config.py [hunk 2]
```diff
@@ -14,6 +15,12 @@
 logger = logging.Logger(__name__)
 
 
+class TagOption(str, Enum):
+    with_lineage = "with_lineage"
+    without_lineage = "without_lineage"
+    skip = "skip"
```
Comment (frsann) [inline metadata-ingestion/src/datahub/ingestion/source/snowflake/snowflake_config.py:21]: I was unsure if it makes sense to use the Snowflake terminology, "lineage", which might get confusing in the Datahub context, or use the term "propagation" which is not the official term, but sometimes used in the documentation. Also the term "tag inheritance" is used sometimes. [Docs](https://docs.snowflake.com/en/user-guide/object-tagging.html#tag-lineage).

Let me know which term makes most sense.

File: metadata-ingestion/src/datahub/ingestion/source/snowflake/snowflake_v2.py [hunk 3]
```diff
@@ -509,10 +531,25 @@ def _process_database(
...
+            if self.config.include_technical_schema:
+                if snowflake_db.tags:
+                    for tag in snowflake_db.tags:
+                        yield from self._process_tag(tag)
+
     def _process_schema(
         self, conn: SnowflakeConnection, snowflake_schema: SnowflakeSchema, db_name: str
     ) -> Iterable[MetadataWorkUnit]:
         schema_name = snowflake_schema.name
+
```
Comment (mayurinehate) [inline metadata-ingestion/src/datahub/ingestion/source/snowflake/snowflake_v2.py:550]: Again, why is this same code repeated below ? The lower one makes more sense as it considers extract_tags config.

File: metadata-ingestion/src/datahub/ingestion/source/snowflake/snowflake_v2.py [hunk 4]
```diff
@@ -674,6 +781,21 @@ def gen_dataset_workunits(
                 view_properties_aspect,
             )
 
+    def gen_tag_workunits(self, tag: SnowflakeTag) -> Iterable[MetadataWorkUnit]:
+        tag_key = str(tag)
+        tag_urn = make_tag_urn(tag_key)
+
+        tag_properties_aspect = TagProperties(
+            name=tag_key,
+            description=f"Represents the Snowflake tag `{tag_key.split('=')[0]}` with value `{tag.value}`.",
```
Comment (mayurinehate) [inline metadata-ingestion/src/datahub/ingestion/source/snowflake/snowflake_v2.py:792]: Maybe add a `from_str` in SnowflakeTag class and use it here to abstract out the tag string formation ?

File: metadata-ingestion/src/datahub/ingestion/source/snowflake/snowflake_v2.py [hunk 5]
```diff
@@ -924,8 +1053,85 @@ def get_views_for_schema(
...
+                        conn, db_name
+                    )
+                )
+
+            if domain == "database":
+                return self.tag_cache.get_database_tags(db_name)
+            elif domain == "schema":
+                assert schema_name is not None
+                tags = self.tag_cache.get_schema_tags(schema_name, db_name)
+            elif domain == "table":  # Views belong to this domain as well.
```
Comment (mayurinehate) [inline metadata-ingestion/src/datahub/ingestion/source/snowflake/snowflake_v2.py:1113]: Nice!

File: metadata-ingestion/src/datahub/ingestion/source/snowflake/snowflake_v2.py [hunk 6]
```diff
@@ -389,6 +408,7 @@ def query(query):
             SourceCapability.LINEAGE_COARSE: "Current role does not have permissions to snowflake account usage views",
             SourceCapability.LINEAGE_FINE: "Current role does not have permissions to snowflake account usage views",
             SourceCapability.USAGE_STATS: "Current role does not have permissions to snowflake account usage views",
+            SourceCapability.TAGS: "Either no tags have been applied to objects or the current role does not have permission to access the objects",
```
Comment (mayurinehate) [inline metadata-ingestion/src/datahub/ingestion/source/snowflake/snowflake_v2.py:411]: Do we need to have any specific access privileges to be able to list and view tags on objects ?

**Review by mayurinehate — CHANGES_REQUESTED** [review]
Thank you for working on this ! 
I have completed the first pass of review. It would be great if you can address the comments. Also, can you please resolve the merge conflicts ?

File: metadata-ingestion/src/datahub/ingestion/source/snowflake/snowflake_v2.py (hunk 6 above)
Comment (frsann) [reply metadata-ingestion/src/datahub/ingestion/source/snowflake/snowflake_v2.py:411]: Per the [documentation](https://docs.snowflake.com/en/user-guide/object-tagging.html#summary-of-ddl-commands-operations-and-privileges):

- the `tag_references` and `tag_references_all_columns`-functions: "Results are only returned for a role that has access to the specified object.". [docs1](https://docs.snowflake.com/en/sql-reference/functions/tag_references.html#usage-notes), [docs2](https://docs.snowflake.com/en/sql-reference/functions/tag_references_all_columns.html). It's my understanding that the role doesn't need explicit access to the information schema, only general USAGE on the database. 
- `snowflake.account_usage.tag_references` is of course part of the account_usage schema, which users are already instructed to grant usage to.

Comment (frsann) [reply metadata-ingestion/src/datahub/ingestion/source/snowflake/snowflake_v2.py:411]: Now that I checked more closely, I noticed the note about the account usage schema in the docs. I'll add tags as one possible use case for that privilege.

**Comment by frsann** [comment]
@mayurinehate  Turns out the single quotes are required in the function calls, so I fixed that. 

For example:
```sql
SELECT tag_database as "TAG_DATABASE",
        tag_schema AS "TAG_SCHEMA",
        tag_name AS "TAG_NAME",
        tag_value AS "TAG_VALUE"
        FROM table(information_schema.tag_references('"my-speci@l_schema"', 'schema'));
```

**Comment by jjoyce0510** [comment]
One thing I want to make sure to call out on this PR.... We've found in the past that if we simply ingest _reference_ to tags (e.g. tag urns for a dataset) without actually _creating_ the tags with DataHub (e.g. by producing the TagProperties or TagKey aspect), then when a user goes to make changes to the Tag via the UI the DataHub app will treat the tag as an entity that does not exist.

We should always consider minting a TagKey or TagProperties (less desirable since it can overwrite a UI-authored description) aspect when ingesting these foreign-authored tags

**Comment by frsann** [comment]
Absolutely! The minting is now behind the same config flag (`include_technical_schema`, defaults to true) as the minting of e.g. tables and the db and schema containers.

File: metadata-ingestion/src/datahub/ingestion/source/snowflake/snowflake_schema.py (hunk 1 above)
Comment (mayurinehate) [reply metadata-ingestion/src/datahub/ingestion/source/snowflake/snowflake_schema.py]: Hey @frsann Let's change this string conversion to use `key:value` form. Also Is there any  explicit need to convert this to lowercase ? I would recommend keeping these to their original case.

File: metadata-ingestion/src/datahub/ingestion/source/snowflake/snowflake_config.py (hunk 2 above)
Comment (mayurinehate) [reply metadata-ingestion/src/datahub/ingestion/source/snowflake/snowflake_config.py:22]: This looks good to me.

File: metadata-ingestion/src/datahub/ingestion/source/snowflake/snowflake_v2.py (hunk 6 above)
Comment (mayurinehate) [reply metadata-ingestion/src/datahub/ingestion/source/snowflake/snowflake_v2.py:416]: We will also need to update the `check_capabilities` method for "Test Connection" feature to mention whether extracting tags is supported for configured role.

File: metadata-ingestion/src/datahub/ingestion/source/snowflake/snowflake_schema.py (hunk 1 above)
Comment (frsann) [reply metadata-ingestion/src/datahub/ingestion/source/snowflake/snowflake_schema.py]: Fixed. I removed the `lower` and replaced it with calling the `self.snowflake_identifier` function which lowercases based on the config.

File: metadata-ingestion/src/datahub/ingestion/source/snowflake/snowflake_v2.py (hunk 6 above)
Comment (frsann) [reply metadata-ingestion/src/datahub/ingestion/source/snowflake/snowflake_v2.py]: Done

File: metadata-ingestion/src/datahub/ingestion/source/snowflake/snowflake_schema.py (hunk 1 above)
Comment (mayurinehate) [reply metadata-ingestion/src/datahub/ingestion/source/snowflake/snowflake_schema.py]: Okay, that seems fair. We should keep display name [tagProperties->name] unlowercased - in exactly same case as snowflake, as we do for snowflake tables, schema, etc.

Comment (mayurinehate) [reply metadata-ingestion/src/datahub/ingestion/source/snowflake/snowflake_schema.py]: Also, I noticed that snowflake tags can have descriptions (for key part)

`create tag cost_center comment = 'cost_center tag';`

Would it make sense to add it to tag description ?

Comment (frsann) [reply metadata-ingestion/src/datahub/ingestion/source/snowflake/snowflake_schema.py]: We could get the comment, but it would of course require an extra `show tags` or `select * from snowflake.account_usage.tags`-query and complicate the tag-workunit creation slightly. It would also give us the `allowed_value` but, it might not be super interesting in Datahub. Let me know if you want it added.   

I fixed the un-lowercasing of the tag name now, only the urn is optionally lowercased.

**Comment by frsann** [comment]
@mayurinehate any chance to get this moving  forward?

**Comment by mayurinehate** [comment]
@frsann I'll get to this early next week. Thank you for your patience.

**Comment by hsheth2** [comment]
@frsann two high-level things before I get into the review

1. My understanding from reading the snowflake docs is that tags in snowflake are global within an account. If that's correct, I'm not confident that the `<db_name>.<schema_name>.<tag_name>=<tag_value>` makes sense. For example, what do we put in the schema_name of a tag is applied to a Snowflake database? I'm thinking that a simple `<optional_configurable_prefix>.<tag_name>=<tag_value>` may make more sense, but definitely want to understand better.
2. It looks like there's a number of merge conflicts between this PR and master. Do you want me to review before or after you've resolved those?

**Comment by frsann** [comment]
1. The tags are defined in a `<database>.<schema>` "namespace".  It's true that the tags are global in the sense that you can apply tag `my_db.public.my_tag` on a table `your_db.staging.some_table`, but you can also create a tag `your_db.public.my_tag` and apply that to the same table without conflicts. 
2. This module seems to be a moving target 😄  I dont mind fixing the conflicts if the review and merge is prompt after that, but I have already made some conflict-fixing commits earlier and I'd prefer to not have to do it many times more.

**Comment by frsann** [comment]
@mayurinehate @hsheth2 the merge conflicts are now fixed. Please have a look.

File: metadata-ingestion/src/datahub/ingestion/source/snowflake/snowflake_config.py [hunk 7]
```diff
@@ -53,6 +60,14 @@ class SnowflakeV2Config(
         default=None, description="Not supported"
     )
 
+    extract_tags: TagOption = Field(
+        default=TagOption.skip,
+        description="""Optionally. Allowed values are `without_lineage`, `with_lineage`, and `skip` (default).
```
Comment (hsheth2) [inline metadata-ingestion/src/datahub/ingestion/source/snowflake/snowflake_config.py:65]:
```suggestion
        description="""Optional. Allowed values are `without_lineage`, `with_lineage`, and `skip` (default).
```

File: metadata-ingestion/src/datahub/ingestion/source/snowflake/snowflake_config.py [hunk 8]
```diff
@@ -76,6 +91,11 @@ def validate_include_column_lineage(cls, v, values):
             )
         return v
 
+    tag_pattern: AllowDenyPattern = Field(
+        default=AllowDenyPattern.allow_all(),
+        description="List of regex patterns for tags to include in ingestion.",
```
Comment (hsheth2) [inline metadata-ingestion/src/datahub/ingestion/source/snowflake/snowflake_config.py:96]:
```suggestion
        description="List of regex patterns for tags to include in ingestion. Only used if `extract_tags` is enabled.",
```

File: metadata-ingestion/src/datahub/ingestion/source/snowflake/snowflake_query.py [hunk 9]
```diff
@@ -105,6 +105,50 @@ def tables_for_schema(schema_name: str, db_name: Optional[str]) -> str:
         and table_type in ('BASE TABLE', 'EXTERNAL TABLE')
         order by table_schema, table_name"""
 
+    @staticmethod
+    def get_all_tags_on_object(
```
Comment (hsheth2) [inline metadata-ingestion/src/datahub/ingestion/source/snowflake/snowflake_query.py:109]: is this without propagation or with?

I think using the _without_propagation or _with_propagation suffixes will help with clarity

File: metadata-ingestion/src/datahub/ingestion/source/snowflake/snowflake_schema.py [hunk 10]
```diff
@@ -86,6 +106,64 @@ class SnowflakeDatabase:
...
+            str, Dict[str, Dict[str, Dict[str, List[SnowflakeTag]]]]
+        ] = defaultdict(
+            lambda: defaultdict(lambda: defaultdict(lambda: defaultdict(list)))
+        )
+
+    def add_database_tag(self, db_name: str, tag: SnowflakeTag) -> None:
+        self._database_tags[db_name].append(tag)
+
+    def get_database_tags(self, db_name: str) -> List[SnowflakeTag]:
+        return self._database_tags.get(db_name, [])
```
Comment (hsheth2) [inline metadata-ingestion/src/datahub/ingestion/source/snowflake/snowflake_schema.py:131]: since we're using a defaultdict, we don't need `.get()` and can just access the field directly right?

File: metadata-ingestion/src/datahub/ingestion/source/snowflake/snowflake_schema.py [hunk 11]
```diff
@@ -86,6 +106,64 @@ class SnowflakeDatabase:
...
+    tags: Optional[List[SnowflakeTag]] = None
+
+
+class _SnowflakeTagCache:
+    def __init__(self) -> None:
+        self._database_tags: Dict[str, List[SnowflakeTag]] = defaultdict(list)
+        self._schema_tags: Dict[str, Dict[str, List[SnowflakeTag]]] = defaultdict(
+            lambda: defaultdict(list)
+        )
+        self._table_tags: Dict[
```
Comment (hsheth2) [inline metadata-ingestion/src/datahub/ingestion/source/snowflake/snowflake_schema.py:118]: some quick comments about structure of these variables would be helpful

e.g. `self._column_tags[database][schema][table_or_view][column] = <list of tags applied to this column>`

File: metadata-ingestion/src/datahub/ingestion/source/snowflake/snowflake_schema.py [hunk 12]
```diff
@@ -358,3 +438,97 @@ def get_fk_constraints_for_schema(
...
+                    snowflake_tag,
+                )
+            else:
+                # This should never happen.
+                self.logger.error(f"Encountered an unexpected domain: {domain}")
+                continue
+
+        return tags
+
+    def get_tags_for_object(
```
Comment (hsheth2) [inline metadata-ingestion/src/datahub/ingestion/source/snowflake/snowflake_schema.py:493]: this is also with propagation right?

File: metadata-ingestion/src/datahub/ingestion/source/snowflake/snowflake_schema.py [hunk 13]
```diff
@@ -358,3 +438,97 @@ def get_fk_constraints_for_schema(
...
+                )
+            )
+        return tags
+
+    def get_tags_on_columns_for_table(
+        self, quoted_table_name: str, db_name: str
+    ) -> Dict[str, List[SnowflakeTag]]:
+        tags: Dict[str, List[SnowflakeTag]] = defaultdict(list)
+        cur = self.query(
+            SnowflakeQuery.get_tags_on_column(db_name, quoted_table_name),
```
Comment (hsheth2) [inline metadata-ingestion/src/datahub/ingestion/source/snowflake/snowflake_schema.py:521]: does this get tags for all columns in the table? if so, can we rename to plural e.g. `get_tags_on_columns`

File: metadata-ingestion/src/datahub/ingestion/source/snowflake/snowflake_tag.py [hunk 14]
```diff
@@ -0,0 +1,139 @@
...
+
+    def _filter_tags(
+        self, tags: Optional[List[SnowflakeTag]]
+    ) -> Optional[List[SnowflakeTag]]:
+        if tags is None:
+            return tags
+
+        allowed_tags = []
+        for tag in tags:
+            tag_identifier = str(tag)
```
Comment (hsheth2) [inline metadata-ingestion/src/datahub/ingestion/source/snowflake/snowflake_tag.py:134]: using str() here feels a bit weird - maybe we can do tag.identifier() or something instead?

File: metadata-ingestion/src/datahub/ingestion/source/snowflake/snowflake_tag.py [hunk 15]
```diff
@@ -0,0 +1,139 @@
...
+        report: SnowflakeV2Report,
+    ) -> None:
+        self.config = config
+        self.data_dictionary = data_dictionary
+        self.report = report
+        self.logger = logger
+
+        self.tag_cache: Optional[_SnowflakeTagCache] = None
+
+    def invalidate_cache(self):
```
Comment (hsheth2) [inline metadata-ingestion/src/datahub/ingestion/source/snowflake/snowflake_tag.py:33]: instead of this invalidate cache method, could we make tag_cache a `dict[db_name, _SnowflakeTagCache]`?

generally cache invalidation bookkeeping is easy to get wrong or miss edge cases, so I'd rather make it obviously correct even if it uses slightly more memory

File: metadata-ingestion/src/datahub/ingestion/source/snowflake/snowflake_v2.py [hunk 16]
```diff
@@ -627,11 +652,20 @@ def _process_database(
                 )
             return
 
+        if self.config.extract_tags is not TagOption.skip:
```
Comment (hsheth2) [inline metadata-ingestion/src/datahub/ingestion/source/snowflake/snowflake_v2.py:655]:
```suggestion
        if self.config.extract_tags != TagOption.skip:
```

**Review by hsheth2 — CHANGES_REQUESTED** [review]
A few minor comments around code cleanup, but overall should be good to merge soon

File: metadata-ingestion/src/datahub/ingestion/source/snowflake/snowflake_query.py [hunk 17]
```diff
@@ -105,6 +105,50 @@ def tables_for_schema(schema_name: str, db_name: Optional[str]) -> str:
...
+        column_name AS "COLUMN_NAME",
+        domain as "DOMAIN"
+        FROM snowflake.account_usage.tag_references
+        WHERE (object_database = '{db_name}' OR object_name = '{db_name}')
+        AND domain in ('DATABASE', 'SCHEMA', 'TABLE', 'COLUMN')
+        AND object_deleted IS NULL;
+        """
+
+    @staticmethod
+    def get_tags_on_columns(db_name: str, quoted_table_identifier: str) -> str:
```
Comment (mayurinehate) [inline metadata-ingestion/src/datahub/ingestion/source/snowflake/snowflake_query.py:141]:
```suggestion
    def get_tags_on_columns_with_propagation(db_name: str, quoted_table_identifier: str) -> str:
```

File: metadata-ingestion/src/datahub/ingestion/source/snowflake/snowflake_schema.py [hunk 18]
```diff
@@ -143,6 +226,8 @@ def get_schemas_for_database(self, db_name: str) -> List[SnowflakeSchema]:
             SnowflakeQuery.schemas_for_database(db_name),
         )
 
+        # get tags as strings
```
Comment (mayurinehate) [inline metadata-ingestion/src/datahub/ingestion/source/snowflake/snowflake_schema.py:229]: Is this comment relevant here ?

File: metadata-ingestion/src/datahub/ingestion/source/snowflake/snowflake_schema.py [hunk 19]
```diff
@@ -358,3 +443,99 @@ def get_fk_constraints_for_schema(
...
+                )
+            )
+        return tags
+
+    def get_tags_on_columns_for_table(
+        self, quoted_table_name: str, db_name: str
+    ) -> Dict[str, List[SnowflakeTag]]:
+        tags: Dict[str, List[SnowflakeTag]] = defaultdict(list)
+        cur = self.query(
+            SnowflakeQuery.get_tags_on_columns(db_name, quoted_table_name),
```
Comment (mayurinehate) [inline metadata-ingestion/src/datahub/ingestion/source/snowflake/snowflake_schema.py:528]:
```suggestion
            SnowflakeQuery.get_tags_on_columns_with_propagation(db_name, quoted_table_name),
```

File: metadata-ingestion/src/datahub/ingestion/source/snowflake/snowflake_schema.py [hunk 20]
```diff
@@ -358,3 +443,99 @@ def get_fk_constraints_for_schema(
...
+
+            # This is the name of the object, unless the object is a column, in which
+            # case the name is in the `COLUMN_NAME` field.
+            object_name = tag["OBJECT_NAME"]
+            # This will be null if the object is a database or schema
+            object_schema = tag["OBJECT_SCHEMA"]
+            # This will be null if the object is a database
+            object_database = tag["OBJECT_DATABASE"]
+
+            domain = tag["DOMAIN"]
```
Comment (mayurinehate) [inline metadata-ingestion/src/datahub/ingestion/source/snowflake/snowflake_schema.py:473]: Can we extend `SnowflakeObjectDomain` from [snowflake/constants.py ](https://github.com/datahub-project/datahub/blob/master/metadata-ingestion/src/datahub/ingestion/source/snowflake/constants.py#L34) to add database and schema  and use it here instead of raw strings ?

File: metadata-ingestion/src/datahub/ingestion/source/snowflake/snowflake_tag.py [hunk 21]
```diff
@@ -0,0 +1,138 @@
...
+    ) -> List[SnowflakeTag]:
+        if self.config.extract_tags == TagOption.without_lineage:
+            if db_name not in self.tag_cache:
+                self.tag_cache[
+                    db_name
+                ] = self.data_dictionary.get_tags_for_database_without_propagation(
+                    db_name
+                )
+
+            if domain == "database":
```
Comment (mayurinehate) [inline metadata-ingestion/src/datahub/ingestion/source/snowflake/snowflake_tag.py:48]: Same comment as above, Let's use `SnowflakeObjectDomain` if possible.

File: metadata-ingestion/src/datahub/ingestion/source/snowflake/snowflake_tag.py [hunk 22]
```diff
@@ -0,0 +1,138 @@
...
+
+    def get_tags_on_object(
+        self,
+        domain: str,
+        db_name: str,
+        schema_name: Optional[str] = None,
+        table_name: Optional[str] = None,
+    ) -> List[SnowflakeTag]:
+        if self.config.extract_tags == TagOption.without_lineage:
+            if db_name not in self.tag_cache:
```
Comment (mayurinehate) [inline metadata-ingestion/src/datahub/ingestion/source/snowflake/snowflake_tag.py:41]: Can we  refractor contents of if condition into another method.  Say -  `get_tags_on_object_without_lineage`
We are trying to move towards small readable and unit-testable methods.

File: metadata-ingestion/src/datahub/ingestion/source/snowflake/snowflake_tag.py [hunk 23]
```diff
@@ -0,0 +1,138 @@
...
+            elif domain == "table":  # Views belong to this domain as well.
+                assert schema_name is not None
+                assert table_name is not None
+                tags = self.tag_cache[db_name].get_table_tags(
+                    table_name, schema_name, db_name
+                )
+            else:
+                raise ValueError(f"Unknown domain {domain}")
+        elif self.config.extract_tags == TagOption.with_lineage:
+            identifier = ""
```
Comment (mayurinehate) [inline metadata-ingestion/src/datahub/ingestion/source/snowflake/snowflake_tag.py:62]: Can we  refractor contents of else condition into another method.  Say -  `get_tags_on_object_with_lineage`

File: metadata-ingestion/src/datahub/ingestion/source/snowflake/snowflake_v2.py [hunk 24]
```diff
@@ -1275,7 +1410,9 @@ def get_columns_for_table(
             )
 
         # Access to table but none of its columns - is this possible ?
-        return columns.get(table_name, [])
+        table_columns = columns.get(table_name, [])
```
Comment (mayurinehate) [inline metadata-ingestion/src/datahub/ingestion/source/snowflake/snowflake_v2.py:1413]: Not sure why this change was needed.

File: metadata-ingestion/src/datahub/ingestion/source/snowflake/snowflake_schema.py [hunk 25]
```diff
@@ -465,16 +469,16 @@ def get_tags_for_database_without_propagation(
...
-            elif domain == "SCHEMA":
+            elif domain == SnowflakeObjectDomain.SCHEMA:
                 tags.add_schema_tag(object_name, object_database, snowflake_tag)
-            elif domain == "TABLE":  # including views
+            elif domain == SnowflakeObjectDomain.TABLE:  # including views
                 tags.add_table_tag(
                     object_name, object_schema, object_database, snowflake_tag
                 )
-            elif domain == "COLUMN":
+            elif domain == SnowflakeObjectDomain.COLUMN:
```
Comment (hsheth2) [inline metadata-ingestion/src/datahub/ingestion/source/snowflake/snowflake_schema.py:481]: nice thanks for refactoring this :)

**Review by hsheth2 — APPROVED** [review]

**Comment by frsann** [comment]
@hsheth2  @mayurinehate  all comments have been fixed, so this PR should not be mergeable
